### PR TITLE
fix: Skip build of Docker image if ImageUri is a valid ECR URL (#2934)

### PR DESCRIPTION
**Pull request by alexisfacques**
#### Which issue(s) does this change fix?

#2934 

#### Why is this change necessary?

- Support deployment of `PackageType: Image` lambda functions of which the underlying Docker image is built/deployed externally to SAM.

#### How does it address the issue?

- Checks whether or not the `ImageUri` property of a `AWS::Serverless::Function`, or `Code.ImageUri` property of a `AWS::Lambda::Function` is a valid ECR URL, based on the resource `PackageType`. If so, skips the local image build and prompts the user with a proper warning, similarly to `Zip` packaged lambdas.

```
  Function:
    Type: AWS::Serverless::Function
    Properties:
      PackageType: Image
      ImageUri: 123456789012.dkr.ecr.eu-west-1.amazonaws.com/function:latest
```

```
% sam build
The resource AWS::Serverless::Function 'Function' has specified ECR registry image for ImageUri. It will not be built and SAM CLI does not support invoking it locally.
```

#### What side effects does this change have?

TBD

#### Checklist

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [x] Write unit tests
- [x] `make pr` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/providers/sam_function_provider.py
Comment (sriram-mv) [inline samcli/lib/providers/sam_function_provider.py:133]: Could we use the enums for Zip and Image, they should be ZIP and IMAGE.

**Review by sriram-mv — CHANGES_REQUESTED** [review]
Generally looks good!

Have some minor comments and once we have tests added, we should be good to go.

**Comment by alexisfacques** [comment]
Changed some unit tests yesterday, unfortunately `make pr` seems broken in Actions. Please let me know If you have any further comments !

**Review by mndeveci — APPROVED** [review]
LGTM, thanks for the contribution and addressing the feedback!

File: samcli/lib/package/ecr_utils.py
Comment (qingchm) [inline samcli/lib/package/ecr_utils.py:9]: nit: add typing for url

**Review by qingchm — APPROVED** [review]

**Review by sriram-mv — APPROVED** [review]

**Comment by qingchm** [comment]
@alexisfacques Hey there, can you please add a fix that lets the make pr passes? Thanks!

**Comment by alexisfacques** [comment]
@qingchm updated and rebased 🙂 

Your `mypy` tests are definitely not messing around when it comes to type hints!
Let me know if I kind be of any further help

Cheers

**Comment by qingchm** [comment]
> @qingchm updated and rebased 🙂
> 
> Your `mypy` tests are definitely not messing around when it comes to type hints!
> Let me know if I kind be of any further help
> 
> Cheers

Thanks a lot for all your timely responses to our comments! Looks good for now! Let me merge it in! I'll let you know later if integ tests have any problems but for now they look really good!

**Comment by marcusletric** [comment]
Hey folks. I'd like to ask for your advice. We've been using sam build in our CI/CD pipelines to build the images. As you introduced this change, the images won't build anymore and later on SAM deploy also fails on CloudFormation.

Maybe I'm missing something here or doing it the wrong way but when I removed the ImageUri parameter on the template then `sam validate` command failed with error that it must be provided.
We also have `PackageType: Image` on our resource, I guess this is why the validation fails.

For now I locked in the sam package version to 1.24.1 on our build env to unblock further releases.

Any suggestions?

**Comment by alexisfacques** [comment]
Heya @marcusletric, I'm very sorry to hear about that, this is an unfortunate regression.

```
AWSTemplateFormatVersion: 2010-09-09

Transform:
  - AWS::Serverless-2016-10-31

Resources:
  HelloWorldFunction1:
    Type: AWS::Serverless::Function
    Properties:
      PackageType: Image
      ImageUri: 123456789012.dkr.ecr.eu-west-1.amazonaws.com/hello_world:latest

  HelloWorldFunction2:
    Type: AWS::Serverless::Function
    Properties:
      PackageType: Image
      ImageUri: helloworldfunction2:python3.8-v1
    Metadata:
      Dockerfile: Dockerfile
      DockerContext: ./hello_world
      DockerTag: python3.8-v1

  HelloWorldFunction3:
    Type: AWS::Serverless::Function
    Properties:
      PackageType: Image
      ImageUri: 123456789012.dkr.ecr.eu-west-1.amazonaws.com/hello_world:latest
    Metadata:
      Dockerfile: Dockerfile
      DockerContext: ./hello_world
      DockerTag: python3.8-v1
```

I believe that your problem relates to `HelloWorldFunction3` of the above, where you had set the Image URI to your target ECR repository; which will in turn skip your Docker image build. 
A quick fix would be to rename the `ImageUri` property to something else (e.g. the local tag of your docker image, this property is overwritten by SAM upon build either way) as shown in `HelloWorldFunction2`; this should pass `sam validate` and allow you to build and deploy your images.

On the long run, SAM checking whether a `DockerContext` resource metadata is set before skipping the build could be a more lasting solution.

**Comment by marcusletric** [comment]
Hey @alexisfacques, thanks for that. I will update our template files.

**Pull request merged**
